### ISSUE TITLE
By default, it seems that learnable parameters are still added to the vision encoder?

### ISSUE THREAD

**Issue by Code-kunkun** [zjysteven/lmms-finetune#11]
As shown in the figure, the trainable parameters of the model are printed as follows.
<img width="714" alt="image" src="https://github.com/user-attachments/assets/b5bcb376-2593-46c1-b124-83bf83aa47ac">

Did I make a mistake somewhere?

**Comment by zjysteven** [comment]
Thanks for reporting this. Could you paste the exact bash commands for launching the training so that we can reproduce and dig into it?

**Comment by Code-kunkun** [comment]
NUM_GPUS=1
DISTRIBUTED_ARGS="
    --nnodes=1 \
    --nproc_per_node ${NUM_GPUS} \
    --rdzv_backend c10d \
    --rdzv_endpoint localhost:4
"

# arguments that are very likely to be changed
# according to your own case
MODEL_ID=llava-interleave-qwen-0.5b                               # model id; pick on by running `python supported_models.py`
TRAIN_DATA_PATH=./example_data/celeba_image_train.json  # path to the training data json file
EVAL_DATA_PATH=./example_data/celeba_image_eval.json    # path to the evaluation data json file (optional)
IMAGE_FOLDER=./example_data/images                      # path to the image root folder; if provided, the image paths in the json should be relative
VIDEO_FOLDER=./example_data/videos                      # path to the video root folder; if provided, the video paths in the json should be relative
DEFAULT_NUM_FRAMES=8                                    # if `num_frames` is not specified in dataset entries, this value will be used to sample frames from videos

USE_LORA=True                                           # whether use lora
Q_LORA=False                                            # whether use q-lora; only effective when `USE_LORA` is True
LORA_R=8                                                # the lora rank
LORA_ALPHA=8                                            # the lora alpha

RUN_ID=${MODEL_ID}_lora-${USE_LORA}_qlora-${Q_LORA}     # a custom run id that determines the checkpoint folder and wandb run name

DS_STAGE=zero3                                          # deepspeed stage; < zero2 | zero3 >
PER_DEVICE_BATCH_SIZE=2                                 # batch size per GPU
GRAD_ACCUM=1                                            # gradient accumulation steps
NUM_EPOCHS=5                                            # number of training epochs

LR=2e-5                                                 # learning rate
MODEL_MAX_LEN=512                                       # maximum input length of the model


torchrun $DISTRIBUTED_ARGS train.py \
    --model_id $MODEL_ID \
    --data_path $TRAIN_DATA_PATH \
    --eval_data_path $EVAL_DATA_PATH \
    --image_folder $IMAGE_FOLDER \
    --video_folder $VIDEO_FOLDER \
    --default_num_frames $DEFAULT_NUM_FRAMES \
    --output_dir ./checkpoints/$RUN_ID \
    --report_to tensorboard \
    --run_name $RUN_ID \
    --deepspeed ./ds_configs/${DS_STAGE}.json \
    --bf16 True \
    --num_train_epochs $NUM_EPOCHS \
    --per_device_train_batch_size $PER_DEVICE_BATCH_SIZE \
    --per_device_eval_batch_size $PER_DEVICE_BATCH_SIZE \
    --gradient_accumulation_steps $GRAD_ACCUM \
    --eval_strategy "epoch" \
    --save_strategy "epoch" \
    --save_total_limit 1 \
    --learning_rate ${LR} \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length $MODEL_MAX_LEN \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --use_lora $USE_LORA \
    --q_lora $Q_LORA \
    --lora_r $LORA_R \
    --lora_alpha $LORA_ALPHA

**Comment by Code-kunkun** [comment]
```
NUM_GPUS=1
DISTRIBUTED_ARGS="
    --nnodes=1 \
    --nproc_per_node ${NUM_GPUS} \
    --rdzv_backend c10d \
    --rdzv_endpoint localhost:4
"

# arguments that are very likely to be changed
# according to your own case
MODEL_ID=llava-interleave-qwen-0.5b                               # model id; pick on by running `python supported_models.py`
TRAIN_DATA_PATH=./example_data/celeba_image_train.json  # path to the training data json file
EVAL_DATA_PATH=./example_data/celeba_image_eval.json    # path to the evaluation data json file (optional)
IMAGE_FOLDER=./example_data/images                      # path to the image root folder; if provided, the image paths in the json should be relative
VIDEO_FOLDER=./example_data/videos                      # path to the video root folder; if provided, the video paths in the json should be relative
DEFAULT_NUM_FRAMES=8                                    # if `num_frames` is not specified in dataset entries, this value will be used to sample frames from videos

USE_LORA=True                                           # whether use lora
Q_LORA=False                                            # whether use q-lora; only effective when `USE_LORA` is True
LORA_R=8                                                # the lora rank
LORA_ALPHA=8                                            # the lora alpha

RUN_ID=${MODEL_ID}_lora-${USE_LORA}_qlora-${Q_LORA}     # a custom run id that determines the checkpoint folder and wandb run name

DS_STAGE=zero3                                          # deepspeed stage; < zero2 | zero3 >
PER_DEVICE_BATCH_SIZE=2                                 # batch size per GPU
GRAD_ACCUM=1                                            # gradient accumulation steps
NUM_EPOCHS=5                                            # number of training epochs

LR=2e-5                                                 # learning rate
MODEL_MAX_LEN=512                                       # maximum input length of the model


torchrun $DISTRIBUTED_ARGS train.py \
    --model_id $MODEL_ID \
    --data_path $TRAIN_DATA_PATH \
    --eval_data_path $EVAL_DATA_PATH \
    --image_folder $IMAGE_FOLDER \
    --video_folder $VIDEO_FOLDER \
    --default_num_frames $DEFAULT_NUM_FRAMES \
    --output_dir ./checkpoints/$RUN_ID \
    --report_to tensorboard \
    --run_name $RUN_ID \
    --deepspeed ./ds_configs/${DS_STAGE}.json \
    --bf16 True \
    --num_train_epochs $NUM_EPOCHS \
    --per_device_train_batch_size $PER_DEVICE_BATCH_SIZE \
    --per_device_eval_batch_size $PER_DEVICE_BATCH_SIZE \
    --gradient_accumulation_steps $GRAD_ACCUM \
    --eval_strategy "epoch" \
    --save_strategy "epoch" \
    --save_total_limit 1 \
    --learning_rate ${LR} \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length $MODEL_MAX_LEN \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --use_lora $USE_LORA \
    --q_lora $Q_LORA \
    --lora_r $LORA_R \
    --lora_alpha $LORA_ALPHA
    
```

**Comment by zjysteven** [comment]
Thanks. Will post updates here.

**Comment by zjysteven** [comment]
I confirm that there is some problem with finding the target modules (partial string keys like "q_proj" are used so the modules in clip's vit will also be included as lora modules). We are actually working on allowing vision encoder to be trained. A fix to this will be made along with those updates. Stay tuned.

**Comment by Code-kunkun** [comment]
So, may I ask how to completely freeze the vision encoder?

**Comment by zjysteven** [comment]
For the time being you can try the code in the `wip-vision-unfreeze` branch, and the `example_scripts/example_image.sh` in that branch (together with the updated code) achieves fully freezing the vision encoder and projector. Although notice that we have only finished tests for llava-1.5 (which should also apply for llava-interleave since there model class/architecture is the same).

**Comment by Code-kunkun** [comment]
Thanks for your quick reply!!!

**Comment by zjysteven** [comment]
The code is merged. Let me know if there are any issues.
